lib/pbench/server/filetree.py

```diff
@@ -0,0 +1,883 @@
+            raise DuplicateDataset(name)
+        if (controller.path / md5_source.name).exists():
+            raise DuplicateDataset(name)
+
+        # Copy the MD5 file first; only if that succeeds, copy the tarball
+        # itself.
+        try:
+            md5_destination = Path(shutil.copy2(md5_source, controller.path))
+        except Exception as e:
+            controller.logger.error(
+                "ERROR copying dataset {} ({}) MD5: {}", name, tarball, e
+            )
+            raise
+
+        try:
+            destination = Path(shutil.copy2(tarball, controller.path))
+        except Exception as e:
+            try:
+                md5_destination.unlink()
+            except Exception as e:
+                controller.logger.error(
+                    "Unable to recover by removing {} MD5 after tarball copy failure: {}",
+                    name,
+                    e,
+                )
+            controller.logger.error(
+                "ERROR copying dataset {} tarball {}: {}", name, tarball, e
+            )
+            raise
+
+        # Restore the SELinux context properly
+        try:
+            if selinux.is_selinux_enabled():
+                selinux.restorecon(destination)
+                selinux.restorecon(md5_destination)
+        except Exception as e:
+            # log it but do not abort
+            controller.logger.error("Unable to set SELINUX context for {}: {}", name, e)
+
+        # To get the new tarball into the server pipeline, we start with a
+        # symlink in the TODO state directory.
+        #
+        # NOTE: there's not much we can do if the link fails, except report it.
+        # The solution is for someone to manually create the link to oil the
+        # tool chain machinery. It doesn't make sense to fail and back out at
+        # this point, and there's no place to report the error aside from the
+        # log.
+        try:
+            controller.link(destination, "TODO")
+        except Exception as e:
+            controller.logger.error(
+                "Failed to link dataset {} into TODO state directory: {}", name, e
+            )
+
+        # If we were able to copy both files, remove the originals
+        try:
+            tarball.unlink()
+            md5_source.unlink()
+        except Exception as e:
+            controller.logger.error("Error removing incoming dataset {}: {}", name, e)
+
+        return Tarball(destination, controller)
+
+    def unpack(self, incoming: Path, results: Path):
+        """
+        Unpack a tarball into the INCOMING directory tree; this assumes that
+        the INCOMING controller directory already exists, which should be
+        ensured by calling this indirectly through the Controller class unpack
+        method.
+
+        TODO: This is a prototype for testing, and doesn't actually unpack
+        the tarball as we're going to be relying on the current unpack pipeline
+        command for some time.
```

Comment:
Ah, I got excited thinking this is replacing our unpack cronjob

Reply:
This was a placeholder that would allow us to pull the bulk of the unpack operation here; but in this PR I haven't tried to spread use of  `FileTree` across the traditional pipeline tools and I decided not to waste too much time filling in all the blanks. It's something to consider as a follow-on. It may make more sense to build the dynamic cache around a modification of this signature allowing specification of a sub-tree or specific file to unpack and cache.